### ISSUE TITLE
Resolve the biggest blockers to Linux building on stable Rust

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-project-goals#116]
| Metadata      | |
| --------      | --- |
| Owner(s)      | @nikomatsakis, @joshtriplett |
| Team(s)       | [compiler](http://github.com/rust-lang/compiler-team), [lang](http://github.com/rust-lang/lang-team), [libs-api](https://www.rust-lang.org/governance/teams) |
| Goal document | [2024h2/rfl_stable](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html) |

## Summary

Stabilize unstable features required by Rust for Linux project including

* Stable support for RFL's customized ARC type
* Labeled goto in inline assembler and extended `offset_of!` support
* RFL on Rust CI
* Pointers to statics in constants

## Tasks and status

* [ ] Overall program management (@nikomatsakis, @joshtriplett)
* [ ] Arbitrary self types v2 (@adetaylor)
  * [x] ~~author RFC~~
  * [x] ~~RFC decision~~ ()
  * [ ] Implementation
  * [ ] Standard reviews ([compiler](http://github.com/rust-lang/compiler-team) ![Team][])
  * [ ] Stabilization decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] Derive smart pointer (@Darksonn)
  * [ ] ~~author RFC~~
  * [x] RFC decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])
  * [ ] Implementation (@dingxiangfei2009)
  * [ ] Author stabilization report (@dingxiangfei2009)
  * [ ] Stabilization decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] `asm_goto` (@nbdd0121)
  * [x] ~~implementation~~
  * [ ] Real-world usage in Linux kernel
  * [ ] Extend to cover full RFC
  * [ ] Author stabilization report
  * [ ] Stabilization decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] Extended `offset_of` syntax (@dingxiangfei2009)
  * [ ] Stabilization report
  * [ ] Stabilization decision ([libs-api](https://www.rust-lang.org/governance/teams) ![Team][])
* [ ] ~~RFL on Rust CI~~ (@Kobzol)
  * [x] ~~implementation~~
  * [ ] Policy draft
  * [ ] Policy decision ([compiler](http://github.com/rust-lang/compiler-team) ![Team][])
* [ ] Pointers to static in constants (@nikomatsakis)
  * [ ] Stabilization report
  * [ ] Stabilization decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])

[Team]: https://img.shields.io/badge/Team%20ask-red

**Comment by nikomatsakis** [comment]
This issue is intended for status updates only.

For general questions or comments, please contact the owner(s) directly.

**Comment by nikomatsakis** [comment]
**Key developments:**

* Implementation work is proceeding in many areas ([arbitrary self types v2](https://github.com/rust-lang/rust/issues/44874#issuecomment-2314739657), [derive smart](https://github.com/rust-lang/rust/pull/125048) [pointer](https://github.com/rust-lang/rust/pull/129467), and [sanitizer support](https://github.com/rust-lang/rust/pull/128348).
* Extended `offset_of` syntax is [stable on nightly!](https://github.com/rust-lang/rust/pull/128284) 🎉  Credit to @dingxiangfei2009 for driving this.
* RFL on Rust CI is implemented but still waiting on documented policy. The first breakage was detected (and fixed) in [#129416](https://github.com/rust-lang/rust/pull/129416). Reading over the PR comments suggests that the policy has not been well communicated to the Rust devs, who weren't clear on who was responsible to fix the breakage etc. Suggested follow-up is to improve the bot message to link to the policy and (as an interim measure) send an email to `all@` and post an Inside Rust blog post.
* [Stabilization report](https://github.com/rust-lang/rust/issues/128183) for Pointers to Statics in Constants is now in FCP. @dingxiangfei2009 to prepare stabilization PR. 💜 
* 

**Next steps**

* [Need to decide on a name for derive-smart pointer](https://github.com/rust-lang/rust/issues/129104). @nikomatsakis to follow-up.
* Experimentation with asm-goto has revealed that a weak point in the design. A typical use case for this feature is to design control-flow, where you have some assembly code that decides which block to branch to. The current design looks like `inline_asm!(..., exit1 = { ... }, exit2 = { ... })`, in which the assembly code can jump to `exit1` or `exit2`. The problem is that, since `inline_asm!` must be in an `unsafe` block, this implies the blocks associated with those labels will be allowed to use unsafe code. To circumvent this, the RFL team is having those blocks simply break from an outer block with a `false` return value, and then branching on that if value outside. That's the same as the C code does, as it happens, but. it's not as nice nor efficient as it could be. Discussion in the meeting suggested that the block bodies should be limited to safe code unless marked as `unsafe`. @joshtriplett to follow-up.
* For Sanitizer support, we will need to stabilize the `-Zfixed-x18` flag as a first step. @Darksonn to start drafting a stabilization report, possibly delegating some questions to @nikomatsakis.

**Comment by nikomatsakis** [comment]
**Key developments:**

* [Kangrejos](https://kangrejos.com), the Rust for Linux Workshop, took place Sep 7-8 in Copenhagen. Session listing and slides are available on <https://kangrejos.com>.
* @Darksonn has opened [RFC #3716][] proposing an approach to manage compiler flags that modify the target ABI and which therefore are only sound if used by all crates. The kernel makes use of a wide number of such flags and it is not practical to have a distinct target for all of the possible combinations. In authoring the RFC it was discovered that there are more such flags than were previously recognized, such as flags that modify LLVM behavior.
* Due to the fact that an increasing number of discussion topics had to do with the compiler, @wesleywiser and @davidtwco from the compiler team have started to join our meetings, which has been helpful in getting faster input from the compiler team.
* The final naming for the "derive-smart-pointer" feature has been settled; it will henceforth be called `#[derive(CoercePointee)]`. @dingxiangfei2009 prepared [PR #131284](https://github.com/rust-lang/rust/pull/131284) to do the rename and is preparing a patch and PR to the rust-for-linux repository. 
* @nbdd0121 authored a [hackmd](https://hackmd.io/@nbdd0121/BJlVrepa0) outlining RFL's experiences with the asm-goto feature and in particular areas for improvement. From this we derived two immediate action items, making the target blocks safe-by-default (https://github.com/rust-lang/rust/issues/119364) and extending `const` to support embedded pointers (https://github.com/rust-lang/rust/issues/128464). 

**Next steps:**

* [RFC #3716][] to be considered by the compiler team.
* Land PR https://github.com/rust-lang/rust/pull/131284 and port code in the Rust-for-Linux repo to use it. Author stabilization report.
* Once FCP completes on https://github.com/rust-lang/rust/issues/132078, review and merge https://github.com/rust-lang/rust/pull/131544 making asm-goto target blocks safe-by-default.
* Decide on extending `const` to support embedded pointers (https://github.com/rust-lang/rust/issues/128464). 
* Stabilization PR for embedding pointers to statics in constants (to be done by @dingxiangfei2009)

[RFC #3716]: https://github.com/rust-lang/rfcs/pull/3716

**Comment by nikomatsakis** [comment]
Brief updates from today's meeting ([minutes](https://hackmd.io/IbiqvBBkRXalMWK0UBSplA)):

* Discussed what goal to set for the next round of project goals. Miguel to prepare a candidate list of stabilization targets, prioritizing language changes (which require widespread edits to the codebase to rebase over) and outright hacks (as is used with rustdoc) and otherwise focusing on "what is needed to get some concrete production use case working". Candidate production use cases we discussed are the android linux kernel configuration or the configuration used by a popular linux distro.
* Discussed synchronization with Debian to see if we can ensure that the version of Rust in their next LTS includes support for Rust 2024, which would make it much easier for the RfL work (along with other crates in the ecosystem) to migrate.
* Compiler team documented policy with respect to RfL on their CI, so that item is done.
* Stabilizationj PR for pointers-to-statics-in-constants (https://github.com/rust-lang/rust/issues/119618) has landed and is stable on nightly, work is needed to update the reference.
* RFC for ABI-relevant compiler flags (https://github.com/rust-lang/rfcs/pull/3716) to be reviewed by compiler team and moved to FCP if appropriate. @wesleywiser to call attention to it in the next compiler team meeting.

**Comment by nikomatsakis** [comment]
Summary of meeting from Nov 20, prepared by @traviscross:

we went through a number of items proposed as 2025H1 project goals. The lang ones all seemed reasonable and achievable project goals. (We didn't get through all the non-lang items, so there's more to cover in the next meeting.)

Wesley mentioned that @Adrian Taylor may not be available to continue pursuing arbitrary self types in 2025, so he's been prioritizing reviews on those to get things in this year, and that we may have to find someone to push this forward if it goes beyond that.

Alice mentioned that CoercePointee may be blocked on arbitrary self types, but wanted to think more about how fundamental that blockage is.

There was a lot of discussion about how to rebuild core in a stable way. Wesley mentioned an interesting thing, which is that the Cargo team would like rustc to support passing a compressed crate file to rustc, and that if we could do that, then we could distribute a core.crate via rustup. Kind of interesting.

There are also some items that the RfL folks want to get into unstable by Rust 1.85 as they may be stuck on that version for a long time. The main one for us is some way to relax the orphan rule. I suggested they file and nominate an ask for a lang experiment here.

**Comment by nikomatsakis** [comment]
Updates:

## 2024H2 cleanup

* Arbitrary self types v2: Stabilization PR is open (rust-lang/rust#135881).
* Stabilize CoercePointee: RFL is using it now; we are making progress towards completing the stabilization.

## 2025H1

* ABI-modifying flags RFC (rust-lang/rfcs#3716) has [completed FCP](https://github.com/rust-lang/rfcs/pull/3716#issuecomment-2614078103) and needs to be merged. Implementation PR [#133138](https://github.com/rust-lang/rust/pull/133138) authored by @azhogin remains under review.
* Other compiler flags, we made a prioritized list. One easy next step would be to rename all `-Z` flags in this list to something stabilizable (e.g., `-C`) that requires `-Zunstable-features`.
  * [ ] `-Zdwarf-version` -- wesley wiser
  * [ ] `-Zdebuginfo-compression`, unblocked
  * [ ] `-Zcrate-attr`, used to configure no-std without requiring it in the source file, no real concerns
  * [ ] `-Zunpretty=expanded`, unblocked, maybe needs a PR that says "don't rely on this", Linux only uses it for debugging macros (i.e. not in the normal build, so it is less critical). Needs a stable name, e.g., `--emit macro-expanded`, and we should make sure output cannot be piped to rustc. `rustfmt` told us (Rust for Linux) that they will try their best to keep `rustfmt` able to format the output of the expansion.
  * [ ] `-Zno-jump-tables`, considered an ABI-modifying flag
  * [ ] `-Zbinary_dep_depinfo=y` -- basically ready to stabilize (@tmandry asked "Do you want toolchain runtimes (libstd, compiler-rt) in your dep info? In my experience this features does 90% of what I want it to do, but removing the inclusion of runtimes is the only question I have before stabilizing", but we don't understand this point yet, as they were not present in the meeting).
* stable rustdoc: PR rust-lang/rust#134531 is under review, should land soon, then next step will be for RFL folks to try it out.
* Clippy configuration: no progress, discussed some new options and posted in thread, very minimal work required here.
* rebuild libcore: @davidtwco wasn't present so no update, but we know progress is being made

### Publicizing this work

We discussed briefly how to better get the word out about this collaboration. Some points:

* @Darksonn will be speaking at Rust Nation
* We could author a Rust-lang blog post, perhaps once the language items are done done done?
* LWN article might be an option

### general-regs-only

We discussed the possibility of a flag to avoid use of floating point registers, no firm conclusion yet reached.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h1

**Comment by nikomatsakis** [comment]
TL;DR: 

We established the precise set of 2025H1 deliverables and we have been tracking them and have begun making progress towards them. Rustdoc has been updated to support extracting doc tests so that the Kernel can execute them in a special environment (this was previously done with a big hack) and RFL is in the process of trying to us that new support. The first PR towards the implementation of [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716) has landed and the ARM team has begun reading early drafts of the design doc for `-Zbuild-core` with the cargo team.

We are also working to finalize the stabilization of the language features that were developed in 2024H2, as two late-breaking complications arose. The first (an interaction between casting of raw pointers and arbitrary self types) is expected to be resolved by limiting the casts of raw pointers, which previously accepted some surprising code. We identified that only a very small set of crates relied on this bug/misfeature; we expect nonetheless to issue a forwards compatibility warning. We are also resolving an issue where `derive(CoercePointee)` was found to reveal the existence of some unstable impls in the stdlib.

**Comment by nikomatsakis** [comment]
Update from our 2025-03-12 meeting ([full minutes](https://hackmd.io/@rust-lang-team/S181TSknyl)):

* RFL team requests someone to look at #138368 which is needed by kernel, @davidtwco to do so.
* `-Zbinary-dep-info` may not be needed; RFL may be able to emulate it.
* rustdoc changes are being incorporated but [@ojeda is exploring whether they can be made more flexible](
https://github.com/rust-lang/rust/pull/134531#discussion_r1894610592).
* [Pre-RFC](https://hackmd.io/@flip1995/By87NXIc1g) authored for clippy stabilization.
* Active iteration on the build-std design; feedback being provided by cargo team.
* @wesleywiser sent a [PR to stabilize `-Zdwarf-version`](https://github.com/rust-lang/rust/pull/136926).
* RfL doesn't use `cfg(no_global_oom_handling)` any more, but it's still in the code for Linux 6.12 which is an LTS kernel. This prevents RfL from using newer versions of rustc. Current plan is to backport a fix to remove usage from LTS kernel.
* Some discussion about best way forward for disabling orphan rule to allow experimentation with no firm conclusion.

**Comment by nikomatsakis** [comment]
TL;DR:

Most of the major items are in an iteration phase. The rustdoc changes for exporting doctests are the furthest along, with a working prototype; the RFL project has been integrating that prototype and providing feedback. Clippy stabilization now has a pre-RFC and there is active iteration towards support for build-std.

Other areas of progress:

* We have an [open PR](https://github.com/rust-lang/rust/pull/136926) to stabilize `-Zdwarf-version`.
* The lang and types team have been discussing the best path forward to resolve [#136702](https://github.com/rust-lang/rust/issues/136702). This is a soundness concern that was raised around certain casts, specifically, casts from a type like `*mut dyn Foo + '_` (with some lifetime) to `*mut dyn Foo + 'static` (with a static lifetime). Rust's defaulting rules mean that the latter is more commonly written with a defaulted lifetime, i.e., just `*mut dyn Foo`, which makes this an easy footgun. This kind of cast has always been dubious, as it disregards the lifetime in a rather subtle way, but when combined with arbitrary self types it permits users to disregard safety invariants making it hard to enforce soundness (see [#136702](https://github.com/rust-lang/rust/issues/136702) for details). The current proposal under discussion in [#136776](https://github.com/rust-lang/rust/issues/136776) is to make this sort of cast a hard error at least outside of an unsafe block; we evaluated the feasibility of doing a future-compatibility-warning and found it was infeasible. Crater runs suggest very limited fallout from this soundness fix but discussion continues about the best set of rules to adopt so as to balance minimizing fallout with overall language simplicity.

**Comment by ojeda** [comment]
Update from our 2025-04-09 meeting ([full minutes](https://hackmd.io/@rust-lang-team/rkqdLEER1l)):

  - Some progress on `arbitrary_self_types`. In particular, decided to do with respect to pin and other related cases.
  
  - `asm_goto` is solved, apart from output operands. For `asm_const`, https://github.com/rust-lang/rust/pull/138618 is nominated.

  - ABI-modifying compiler flags: some PRs waiting review, e.g. https://github.com/rust-lang/rust/pull/138736.

  - `--crate-attr` RFC is up: https://github.com/rust-lang/rfcs/pull/3791.

  - `-Zsanitize-kcfi-arity`'s implementation PR got merged: https://github.com/rust-lang/rust/pull/138368. If all is good from the Linux side, a stabilization PR will be sent.

  - CFI `core::fmt` issue: https://github.com/rust-lang/rust/issues/115199.

  - Discussion around `bindgen`, `repr(align)` and packed types. RFC nominated for lang discussion: https://github.com/rust-lang/rfcs/pull/3718#issuecomment-2790654254.

**Comment by ojeda** [comment]
Update from our 2025-05-07 meeting ([full minutes](https://hackmd.io/@rust-lang-team/S1Y3l7Kxel)):

  - Enthusiasm and plans for RustWeek.

  - `arbitrary_self_types`: update from @dingxiangfei2009 at https://rust-lang.zulipchat.com/#narrow/channel/425075-rust-for-linux/topic/2025-05-07.20meeting/near/516734641 -- he plans to talk to types in order to find a solution. @davidtwco will ping @petrochenkov about `rustc_resolve`.

  - Sanitizer support and `#[sanitize(off)]`: discussed by lang at https://github.com/rust-lang/rust/pull/123617#issuecomment-2859621119. Discussion about allowing to disable particular sanitizers. Older concern from compiler at https://github.com/rust-lang/rust/pull/123617#issuecomment-2192330122.

  - `asm_const` with pointers support: lang talked about it -- lang will want an RFC: https://github.com/rust-lang/rust/issues/128464#issuecomment-2861515372.

  - ABI-modifying compiler flags: two MCPs filled: https://github.com/rust-lang/compiler-team/issues/868 (`-Zretpoline` and `-Zretpoline-external-thunk`) and https://github.com/rust-lang/compiler-team/issues/869 (`-Zharden-sls`).

    Implementation PR for `-Zindirect-branch-cs-prefix` at https://github.com/rust-lang/rust/pull/140740 that goes on top of https://github.com/rust-lang/rust/pull/135927.

    @davidtwco agreed there is likely no need for a separate MCP for this last one, i.e. it could go into the `-Zretpoline*` one. @azhogin pinged about this at https://github.com/rust-lang/rust/pull/135927#issuecomment-2859906060.
  
  - `--crate-attr`: @Mark-Simulacrum was pinged and he is OK to adopt the RFC (https://github.com/rust-lang/rfcs/pull/3791).

**Comment by nikomatsakis** [comment]
TL;DR:

May saw significant progress on compiler flags, with MCPs for `-Zharden-sls` and `-Zretpoline*` being accepted. Several PRs were in progress (#135927, #140733, #140740) that could potentially be combined, with the implementation approach matching clang's flag naming conventions for consistency. The RFC for configuring no-std externally (#3791) entered T-compiler FCP with positive signals, and build-std discussions at the All Hands produced some consensus between libs and compiler teams, though more Cargo team involvement was needed.

The Rust for Linux team had strong participation at Rust Week, with many team members attending (Alice, Benno, Björn, Boqun, Gary, Miguel, Trevor). During the All Hands, attendees participated in a fun exercise predicting what percentage of the kernel will be written in Rust by 2035 - currently only about 0.1% of the kernel's 40M total lines are in Rust.

On language features, during May we continued work on arbitrary self types v2, where Ding focused on resolving the dichotomy between `Deref::Target` vs `Receiver::Target`. One potential approach discussed was splitting the feature gate to allow arbitrary self types only for types implementing `Deref`, which would cover the kernel use case. For `derive(CoercePointee)`, we continued waiting on PRs #136764 and #136776, with the latter needing diagnostic work.

The All Hands meeting also produced interesting developments on field projections, with Benno working on an approach that reuses borrow checker logic to extend what we do for `&` and `&mut` to custom types using the `->` syntax. Alice also presented a new proposal for AFIDT/RPITIDT and placement ([discussed here](https://rust-lang.zulipchat.com/#narrow/channel/486433-all-hands-2025/topic/Placement.20new.20in.20Rust/near/518803523)).

**Comment by tomassedovic** [comment]
## In-place initialization ##

Ding opened a [PR#142518](https://github.com/rust-lang/rust/pull/142518) that implements the [in-place initialization experiment](https://github.com/rust-lang/lang-team/issues/336).

## `arbitrary_self_types`

Ding is working on an experimental implementation ([PR#143527](https://github.com/rust-lang/rust/pull/143527)).

## Queries on GCC-style inline assembly statements:

Ding opened a PR to Clang (a C frontend for LLVM): https://github.com/llvm/llvm-project/pull/143424 and got it merged.

This is part of the LLVM/Clang issues the Rust for Linux project needs: https://github.com/Rust-for-Linux/linux/issues/1132.

## `-Zindirect-branch-cs-prefix`:

We've discussed whether this needs to be a separate target feature vs. a modifier on the existing `retpoline` one. Josh argued that since having this enabled without retpoline doesn't make sense, it should be a modifier. On the other hand, Miguel mentioned having a separate feature would be easier to implement.

Ultimately, this is a compiler question and should be resolved here: https://github.com/rust-lang/rust/pull/140740

Note: `-Cmin-function-alignment` will be another similar case.

## Stabilizing `AddressSanitizer` and `LeakSanitizer`:

* https://github.com/rust-lang/rust/pull/123617
* https://github.com/rust-lang/rust/pull/142681

In light of the newly-proposed `#[sanitize(xyz = "on|off")]` syntax, we've discussed whether it makes sense to add a shorthand to enable/disable all of them at once (e.g. `#[sanitize(all = "on|off")]`). The experience from the field suggests that this is rarely something people do.

We've also discussed what values should the options have (e.g. `"yes"`/`"no"` vs. `"on"`/`"off"` or `true`/`false`). No strong preferences, but in case of an error, the compiler should suggest the correct value to use.

P.S.: There will be a Lang design meeting regarding in-place initialization on Wednesday 2025-07-30: https://github.com/rust-lang/lang-team/issues/332.

**Comment by tomassedovic** [comment]
## Compiler flags

- [Add -Zindirect-branch-cs-prefix PR#140740](https://github.com/rust-lang/rust/pull/140740) -- landed
- [Fix -Zregparm for LLVM builtins PR#145309](https://github.com/rust-lang/rust/pull/145309) -- also landed
- [-Zharden-sls PR#136597](https://github.com/rust-lang/rust/pull/136597) -- ready for reviews
- [Sanitizers target modificators PR#138736](https://github.com/rust-lang/rust/pull/138736) -- close to landing
- [Add assembly test for -Zreg-struct-return option PR#145382](https://github.com/rust-lang/rust/pull/145382) -- landed

## `arbitrary_self_types~ and `derive(CoercePointee)`

@dingxiangfei2009 is working on splitting the `Deref` and `Receiver` traits (the original plan was to make `Deref` a subtrait of `Receiver`). Ding has investigated and [written down the method probing algorithm](https://hackmd.io/@rust-for-linux-/rySktshdex) and documented the changes he's proposing. He'll be work on a PR that could show this in action next.

## Support for pointers with `asm_const`

We've scheduled a design meeting with the Lang team to read through [@Darksonn's RFC](https://github.com/rust-lang/rfcs/pull/3848). It is [scheduled for 2025-10-08](https://github.com/rust-lang/lang-team/issues/347).

## Field projections

@BennoLossin and @dingxiangfei2009 put together an experimental implementation of field projection: https://github.com/BennoLossin/rust/tree/field-projections.

They plan to polish this up, cover the missing cases and then post a PR. @tmandry's been championing this from the Lang side and trying to burn down some of the open questions.

## In-place initialization

Experimental PR: https://github.com/rust-lang/rust/pull/142518

Progress has slowed somewhat, it's not clear which of the proposed in-place init options to go forward with. Alice, Ding and Benno will talk and see if they can get this unblocked.

## Reducing codegen size

A discussion on the status / next steps towards stabilizing cfg(no_fp_fmt_parse) [PR#86048](https://github.com/rust-lang/rust/pull/86048) triggered a broader discussion of Libs team's concern for littering the standard libraries with endless `cfg` flags and other ways of approaching this. For example, getting to a more minimal `core` library that could be produced using build-std and then implement the missing capabilities in the kernel.

This discussion suggested multiple avenues that I plan to follow: checking back with Libs on their current stance of `cfg`, appetite for building a smaller core, specifying behaviours through custom targets, etc.

## NUL-terminated file names with #[track_caller] aka `file_with_nul`

The Libs-API team discussed this on Tuesday 2025-08-26 and agreed to the feature, but under a different name: the function will be called: `file_as_c_str` instead of `file_with_nul`:

https://github.com/rust-lang/rust/issues/141727

The tracking issue now has a new FCP proposed with that name change.

In the meantime, @Darksonn opened a  PR renaming the function in nightly: https://github.com/rust-lang/rust/pull/145928.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by tomassedovic** [comment]
Update from the 2025-11-05 meeting.

## `Deref`/`Receiver`

@dingxiangfei2009 posted his reasoning for the trait split in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/522311-t-lang.2Fcustom-refs/topic/Consequences.20of.20making.20Deref.20a.20subtrait.20of.20Receiver/with/547014978) and suggested adding a second RFC to explain.

@traviscross recommended writing a Reference PR. The style forces one to explain the model clearly which should then make writing the RFC easier.

The lang experiment PR for arbitrary self types have feature gates for the two options we're exploring.

## Arbitrary Self Types and `derive(CoercePointee)` / [tracking issue #44874](https://github.com/rust-lang/rust/issues/44874)

theemathas opened an issue [`derive(CoercePointee)` accepts `?Sized + Sized` #148399](https://github.com/rust-lang/rust/issues/148399). This isn't a critical issue, just an error that arguably should be a lint.

@BoxyUwU opened a fix for a `derive(CoercePointee)` blocker: [Forbid freely casting lifetime bounds of dyn-types
](https://github.com/rust-lang/rust/pull/136776).

## [RFC #3851: Supertrait Auto-impl](https://github.com/rust-lang/rfcs/pull/3851)

@dingxiangfei2009 is working on the implementation (the parser and HIR interface for it). Ding's also working on a more complete section dedicated to [questions raised by obi1kenobi](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2430827437)

## Field projections

@BennoLossin has been posting super detailed updates on [the tracking issue](https://github.com/rust-lang/rust-project-goals/issues/390#issuecomment-3438476150)

We've discussed the idea of virtual places (see [Zulip thread where they were proposed](https://rust-lang.zulipchat.com/#narrow/channel/522311-t-lang.2Fcustom-refs/topic/Field.20projections.20and.20places/with/553831123)).

## Inlining C code into Rust code

Matt Mauer had an idea to compile C code into LLVM bytecode (instead of object file) and then the llvm-link tool to merge them together and treat everything in the second bytecode file as a static inlined function. Matt suggested we could integrate this into the rustc passes.

This would make it easy to inline certain functions into Rust code without full LTO.

Relevant [Zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/pre-MCP.20vibe.20check.3A.20-Zinternalize-bitcode/near/546294314).

This sounds like a good candidate for the next Project Goals period.

**Comment by tomassedovic** [comment]
Update from the 2025-12-03 meeting.

## `Deref` / `Receiver`

Ding keeps working on the Reference draft. The idea is still not well-proliferated and people are not convinced this is a good way to go. We hope the method-probing section in Reference PR could clear thins up.

We're keeping the supertrait auto-impl experiment as an alternative.

## [RFC #3851: Supertrait Auto-impl](https://github.com/rust-lang/rfcs/pull/3851)

Ding addressed Predrag's requests on SemVer compatibility. He's also opened an implementation PR: https://github.com/rust-lang/rust/pull/149335. Here's the tracking issue: https://github.com/rust-lang/rust/issues/149556.

## `derive(CoercePointee)`

Ding opened a PR to require additional checks for DispatchFromDyn: https://github.com/rust-lang/rust/pull/149068

## In-place initialization

Ding will prepare material for a discussion at the LPC (Linux Plumbers Conference). We're looking to hear feedback on the end-user syntax for it.

The feature is going quite large, Ding will check with Tyler on the whether this might need a series of RFCs.

The various proposals on the table continue being discussed and there are signs (albeit slow) of convergence. The placing function and guaranteed return ones are superseded by outpointer. The more ergonomic ideas can be built on top. The guaranteed value placement one would be valuable in the compiler regardless and we're waiting for Olivier to refine it.

The feeling is that we've now clarified the constraints that the proposals must operate under.

## Field projections

Nadri's Custom places proposal is looking good at least for the user-facing bits, but the whole thing is growing into a large undertaking. Benno's been focused on academic work that's getting wrapped up soon. The two will sync afterwards.

**Comment by tomassedovic** [comment]
Quick bit of great news: Rust in the Linux kernel is no longer treated as an experiment, it's here to stay 🎉 

https://lwn.net/SubscriberLink/1050174/63aa7da43214c3ce/

**Comment by tomassedovic** [comment]
Update from the 2026-04-08 meeting:

## zerocopy features in Rust's std

zerocopy uses two traits that are both polyfills for unstable traits : `KnownLayout` (for `ptr_metadata`) and `Immutable` (for `Freeze`). It would [help maintenance of zerocopy](https://github.com/rust-lang/rust/issues/81513#issuecomment-2414679019) (which Rust for Linux plans to start using) if these were stabilised.

`ptr_metadata` is something the team wants in the kernel independently. It's possibly blocked on (or at least might have interactions with) the [Sized Hierarchy work](https://github.com/rust-lang/rust/issues/144404).

`Freeze` (now `NoCell`) has an RFC here: https://github.com/rust-lang/rfcs/pull/3633.

## `Deref`/`Receiver`

@jackh726 started reviewing @dingxiangfei2009's [rust#146095 split the autoderef chain PR](https://github.com/rust-lang/rust/pull/146095) and feels it's not ready to go in front of the full Lang team.

We also discussed the dependence/independence of the `Deref` and `Receiver` implementations, in particular whether it ever makes sense to implement `Deref` but *not* `Receiver`. @joshtriplett suggested gathering examples for cases like that (where you can't use the type as a `Self` type in the function declaration, but allow calling methods on it).

The current plan for the experiment is to have these traits separate, but have the compiler enforce that if they implement the same type, their targets are identical. This will let us open the door for any future possibilities (a supertrait / subtrait relation, or having diverging targets in the future).

We want to experiment to see where and how these traits and their possible evolution might be helpful.

## null-ptr-deref

The team would like to have a (an optional) compiler guarantee, that the compiler never removes null checks on raw pointers. What can currently happen in C is that if you deref a null pointer, the compiler can do optimisations including removing any subsequent checks whether that pointer is null, because dereferencing a null pointer is undefined behaviour.

But the null check can still help prevent further bugs and in C, the kernel now disables the optimisation that would remove it.

@ojeda is going to open an MCP for this.

## In-Place Initialization

@BennoLossin opened a proposal for an in-person room at the 2026 All Hands for In-place initialization: https://github.com/rust-lang/all-hands-2026/issues/17.

Here's a meta issue tracking all the proposals and discussions about the feature: https://github.com/rust-lang/rust/issues/153825.

The design space is complex and the team hopes that discussing it in person will help move it forward.

**Comment by nxsaken** [comment]
The 2025h2 goal period is now over, so I'm closing this issue. Thanks to everyone who contributed! This work continues in the 2026 period via the [Rust for Linux roadmap](https://rust-lang.github.io/rust-project-goals/2026/roadmap-rust-for-linux.html), which includes the [Arbitrary Self Types](https://github.com/rust-lang/rust-project-goals/issues/619), [build-std](https://github.com/rust-lang/rust-project-goals/issues/274), [Full Const Generics](https://github.com/rust-lang/rust-project-goals/issues/100), [Const Traits](https://github.com/rust-lang/rust-project-goals/issues/106), [Field Projections](https://github.com/rust-lang/rust-project-goals/issues/390), [In-place initialization](https://github.com/rust-lang/rust-project-goals/issues/395), [Immobile types and guaranteed destructors](https://github.com/rust-lang/rust-project-goals/issues/635), and [Supertrait `auto impl`](https://github.com/rust-lang/rust-project-goals/issues/636) goals. Authors are welcome to leave a final comment about the goal's outcome.

**Comment by ojeda** [comment]
There has been a lot of progress and the regular meetings were as useful as always, thanks a lot everyone!

(and to Tomas for keeping track of things and taking minutes, even when the meetings ran over :)

I hope the next period will be the one that sees the Linux kernel essentially removing the last three unstable language features we use (modulo the new ones being added), and getting very close to finishing the list of the compiler features side too.
